docs/core/modules.rst

```diff
@@ -0,0 +1,132 @@
+.. core_modules:
+
+Modules
+=======
+
+.. module:: superdesk.core
+
+Loading Modules
+---------------
+
+The :class:`Module <module.Module>` class is used to register modules with the system, and provides attributes
+used to define certain behaviour of the module.
+
+An example of a module's root file::
+
+    # file: my/module/__init__py
+    from superdesk.core.module import Module, SuperdeskAsyncApp
+
+    def init(app: SuperdeskAsyncApp):
+        ...
+
+    module = Module(
+        name="my.module",
+        init=init,
+        frozen=True,
+        priority=100
+    )
+
+Then you would add the following to your :ref:`settings.modules` config::
+
+    # file: settings.py
+    MODULES = ["my.module"]
+
+The system would then load this module when the application starts
+
+Module Loading Sequence
+-----------------------
+
+The :class:`SuperdeskAsyncApp <superdesk.core.app.SuperdeskAsyncApp>` app is in charge of loading modules,
+when the `start` function is run on the app.
+
+Modules are loaded in the following 3 stages.
+
+1. Import python module
+2. Load module local configs
+3. Run module init functions
+
+The app will import ALL modules first, then ALL module configs are loaded, then ALL module init functions are run.
+This way we can ensure that all python modules and configs are loaded **before** initialising the modules themself.
+
+Module Configuration
+--------------------
+
+A module can optionally have a :ref:`config <core_config>` loaded on app startup. This can be achieved by setting config options on the Module
+instance.::
+
+    # file: my/module/__init__py
+    from typing import Optional, Dict, Any
+
+    from superdesk.core.module import Module, SuperdeskAsyncApp
+    from superdesk.core.config import ConfigModel
+
+
+    # Create a new Config class that inherits from `ConfigModel`
+    class ModuleConfig(ConfigModel):
+        # Make sure to provide defaults for all fields
+        # otherwise validation would fail
+        uri: str = "localhost"
+        port: int = 27017
+        options: Optional[Dict[str, Any]] = None
+
+    # Create a module local instance of the class,
+    # to be used at runtime
+    config = ModuleConfig()
```

Comment:
I'm wondering if it wouldn't be better usage to call it with app so when it's initialized it will actually have the values already populated, eg.:
```python3
app = get_current_app()
config = ModuleConfig(app)
```
otherwise this could in theory lead to usage without proper initialization.

Reply:
I'd rather have a direct dependence of App for Config. At the moment it's pure python/pedantic.

App is only available after app has been built.

We can place in a function on the ConfigModel class that does the loading, and sets an attribute saying it's loaded. Then when trying to access a variable, we can throw an exception if it's not loaded.

Reply:
I think it should actually depend on it rather explicitly than implicitly. Using the config before the app was loaded is always wrong anyhow, isn't it?

Reply:
but might be more complicated to test different configs

Reply:
I see